File: distribute-candies/distribute-candies.js
/**
 * @param {number[]} candies
 * @return {number}
 */
var distributeCandies = function (candies) {
  let p = new Map()

  for (let item of candies) {
    p.set(item, ~~p.get(item) + 1)
  }

  let res = [...p.values()].sort((a, b) => a - b)

  return Math.min(res.length, candies.length >> 1)
};

var distributeCandies2 = function(candies) {
  return Math.min(candies.length / 2, [...new Set(candies)].length)
};

File: leetcode/distribute-candies/distribute-candies.py
class Solution:
    def distributeCandies(self, candies):
        """
        :type candies: List[int]
        :rtype: int
        """
        d = {}
        for item in candies:
            if item in d:
                d[item] += 1
            else:
                d[item] = 1

        res = sorted(list(d.values()))

        return min(len(res), len(candies) >> 1)

    def distributeCandies2(self, candies):
        a = len(list(set(candies)))
        b = len(candies)
        if a >= b / 2:
            return b / 2
        return a
